Failed: JdbcProjectionTest exactlyOnceShouldRestartFromPreviousOffset

```
<-- [akka.projection.jdbc.JdbcProjectionTest: exactlyOnceShouldRestartFromPreviousOffset] End of log messages of test that failed with Expected exception
[info] - exactlyOnceShouldRestartFromPreviousOffset *** FAILED ***
[info]   java.lang.AssertionError: Expected exception
[info]   at org.junit.Assert.fail(Assert.java:88)
[info]   at akka.projection.jdbc.JdbcProjectionTest.exactlyOnceShouldRestartFromPreviousOffset(JdbcProjectionTest.java:259)
```
https://travis-ci.com/github/akka/akka-projection/jobs/371418704#L564

This may have something to do with non-integration tests not running in parallel.  I debated whether the H2 tests should belong in integration tests config or not, and opted to keep them as unit tests since they're fast.  It may be worth experimenting if `parallelExecution := false` is the issue here.

--- a/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
+++ b/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
@@ -10,6 +10,7 @@ import akka.actor.testkit.typed.javadsl.LogCapturing;
 import akka.actor.testkit.typed.javadsl.TestKitJunitResource;
 import akka.actor.testkit.typed.javadsl.TestProbe;
 import akka.japi.function.Function;
+import akka.japi.pf.Match;
 import akka.projection.Projection;
 import akka.projection.ProjectionContext;
 import akka.projection.ProjectionId;
@@ -180,12 +181,16 @@ public class JdbcProjectionTest extends JUnitSuite {
     return "fail on envelope with offset: [" + offset + "]";
   }
 
-  private PartialFunction<TestSubscriber.SubscriberEvent, Object> expectErrorMessage(String msg) {
-    return akka.japi.pf.Match.<TestSubscriber.SubscriberEvent, Object, TestSubscriber.OnError>match(
-            TestSubscriber.OnError.class,
-            err -> err.cause().getMessage().equals(msg),
-            event -> null)
-        .build();
+  private void expectNextUntilErrorMessage(TestSubscriber.Probe<Done> probe, String msg) {
+    probe.request(1);
+    PartialFunction<TestSubscriber.SubscriberEvent, Boolean> pf =
+        Match.<TestSubscriber.SubscriberEvent, Boolean, TestSubscriber.OnError>match(
+                TestSubscriber.OnError.class,
+                err -> err.cause().getMessage().equals(msg),
+                event -> true)
+            .match(TestSubscriber.OnNext.class, event -> false)
+            .build();
+    if (!probe.expectEventPF(pf)) expectNextUntilErrorMessage(probe, msg);
   }
 
   private JdbcHandler<Envelope, PureJdbcSession> concatHandler(StringBuffer str) {
@@ -273,8 +278,7 @@ public class JdbcProjectionTest extends JUnitSuite {
           probe.request(3);
           probe.expectNextN(3);
           assertEquals("abc|def|ghi|", str.toString());
-          probe.request(1);
-          probe.expectEventPF(expectErrorMessage(failMessage(4)));
+          expectNextUntilErrorMessage(probe, failMessage(4));
         });
   }
 
@@ -326,8 +330,7 @@ public class JdbcProjectionTest extends JUnitSuite {
           probe.request(2);
           probe.expectNextN(2);
           assertEquals("abc|def|ghi|", str.toString());
-          probe.request(1);
-          probe.expectEventPF(expectErrorMessage(failMessage(4)));
+          expectNextUntilErrorMessage(probe, failMessage(4));
         });
 
     assertStoredOffset(projectionId, 3L);

--- a/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
+++ b/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
@@ -9,6 +9,7 @@ import akka.NotUsed;
 import akka.actor.testkit.typed.javadsl.LogCapturing;
 import akka.actor.testkit.typed.javadsl.TestKitJunitResource;
 import akka.actor.testkit.typed.javadsl.TestProbe;
+import akka.japi.JavaPartialFunction;
 import akka.japi.function.Function;
 import akka.projection.Projection;
 import akka.projection.ProjectionContext;
@@ -20,10 +21,13 @@ import akka.projection.jdbc.javadsl.JdbcHandler;
 import akka.projection.jdbc.javadsl.JdbcProjection;
 import akka.projection.testkit.javadsl.TestSourceProvider;
 import akka.projection.testkit.javadsl.ProjectionTestKit;
+import akka.stream.impl.ActorSubscriberMessage;
 import akka.stream.javadsl.FlowWithContext;
 import akka.stream.javadsl.Source;
+import akka.stream.testkit.TestSubscriber;
 import com.typesafe.config.Config;
 import com.typesafe.config.ConfigFactory;
+import org.jetbrains.annotations.NotNull;
 import org.junit.*;
 import org.scalatestplus.junit.JUnitSuite;
 import scala.Option;
@@ -175,6 +179,20 @@ public class JdbcProjectionTest extends JUnitSuite {
     return "fail on envelope with offset: [" + offset + "]";
   }
 
+  @NotNull
+  private JavaPartialFunction<TestSubscriber.SubscriberEvent, Object> expectErrorMessage(
+      String msg) {
+    return new JavaPartialFunction<>() {
+      public String apply(TestSubscriber.SubscriberEvent in, boolean isCheck) {
+        if (in instanceof TestSubscriber.OnError) {
+          TestSubscriber.OnError err = (TestSubscriber.OnError) in;
+          if (err.cause().getMessage().equals(msg)) return null;
+        }
+        throw noMatch();
+      }
+    };
+  }
+
   private JdbcHandler<Envelope, PureJdbcSession> concatHandler(StringBuffer str) {
     return concatHandler(str, __ -> false);
   }
@@ -250,16 +268,19 @@ public class JdbcProjectionTest extends JUnitSuite {
             projectionId,
             sourceProvider(entityId),
             jdbcSessionCreator,
-            // fail on forth offset
+            // fail on fourth offset
             () -> concatHandler(str, offset -> offset == 4),
             testKit.system());
 
-    try {
-      projectionTestKit.run(projection, () -> assertEquals("abc|def|ghi|", str.toString()));
-      Assert.fail("Expected exception");
-    } catch (RuntimeException e) {
-      assertEquals(failMessage(4), e.getMessage());
-    }
+    projectionTestKit.runWithTestSink(
+        projection,
+        (probe) -> {
+          probe.request(3);
+          probe.expectNextN(3);
+          assertEquals("abc|def|ghi|", str.toString());
+          probe.request(1);
+          probe.expectEventPF(expectErrorMessage(failMessage(4)));
+        });
   }
 
   @Test
@@ -299,21 +320,20 @@ public class JdbcProjectionTest extends JUnitSuite {
                 projectionId,
                 sourceProvider(entityId),
                 jdbcSessionCreator,
-                // fail on forth offset
+                // fail on fourth offset
                 () -> concatHandler(str, offset -> offset == 4),
                 testKit.system())
             .withSaveOffset(1, Duration.ZERO);
 
-    try {
-      projectionTestKit.run(
-          projection,
-          () -> {
-            assertEquals("abc|def|ghi|", str.toString());
-          });
-      Assert.fail("Expected exception");
-    } catch (RuntimeException e) {
-      assertEquals(failMessage(4), e.getMessage());
-    }
+    projectionTestKit.runWithTestSink(
+        projection,
+        (probe) -> {
+          probe.request(2);
+          probe.expectNextN(2);
+          assertEquals("abc|def|ghi|", str.toString());
+          probe.request(1);
+          probe.expectEventPF(expectErrorMessage(failMessage(4)));
+        });
 
     assertStoredOffset(projectionId, 3L);
 

--- a/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
+++ b/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
@@ -9,7 +9,6 @@ import akka.NotUsed;
 import akka.actor.testkit.typed.javadsl.LogCapturing;
 import akka.actor.testkit.typed.javadsl.TestKitJunitResource;
 import akka.actor.testkit.typed.javadsl.TestProbe;
-import akka.japi.JavaPartialFunction;
 import akka.japi.function.Function;
 import akka.projection.Projection;
 import akka.projection.ProjectionContext;
@@ -19,18 +18,20 @@ import akka.projection.jdbc.internal.JdbcOffsetStore;
 import akka.projection.jdbc.internal.JdbcSettings;
 import akka.projection.jdbc.javadsl.JdbcHandler;
 import akka.projection.jdbc.javadsl.JdbcProjection;
-import akka.projection.testkit.javadsl.TestSourceProvider;
 import akka.projection.testkit.javadsl.ProjectionTestKit;
-import akka.stream.impl.ActorSubscriberMessage;
+import akka.projection.testkit.javadsl.TestSourceProvider;
 import akka.stream.javadsl.FlowWithContext;
 import akka.stream.javadsl.Source;
 import akka.stream.testkit.TestSubscriber;
 import com.typesafe.config.Config;
 import com.typesafe.config.ConfigFactory;
-import org.jetbrains.annotations.NotNull;
-import org.junit.*;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
 import org.scalatestplus.junit.JUnitSuite;
 import scala.Option;
+import scala.PartialFunction;
 import scala.concurrent.Await;
 import scala.concurrent.Future;
 
@@ -179,18 +180,12 @@ public class JdbcProjectionTest extends JUnitSuite {
     return "fail on envelope with offset: [" + offset + "]";
   }
 
-  @NotNull
-  private JavaPartialFunction<TestSubscriber.SubscriberEvent, Object> expectErrorMessage(
-      String msg) {
-    return new JavaPartialFunction<TestSubscriber.SubscriberEvent, Object>() {
-      public String apply(TestSubscriber.SubscriberEvent in, boolean isCheck) {
-        if (in instanceof TestSubscriber.OnError) {
-          TestSubscriber.OnError err = (TestSubscriber.OnError) in;
-          if (err.cause().getMessage().equals(msg)) return null;
-        }
-        throw noMatch();
-      }
-    };
+  private PartialFunction<TestSubscriber.SubscriberEvent, Object> expectErrorMessage(String msg) {
+    return akka.japi.pf.Match.<TestSubscriber.SubscriberEvent, Object, TestSubscriber.OnError>match(
+            TestSubscriber.OnError.class,
+            err -> err.cause().getMessage().equals(msg),
+            event -> null)
+        .build();
   }
 
   private JdbcHandler<Envelope, PureJdbcSession> concatHandler(StringBuffer str) {

--- a/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
+++ b/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
@@ -327,6 +327,12 @@ public class JdbcProjectionTest extends JUnitSuite {
     projectionTestKit.runWithTestSink(
         projection,
         (probe) -> {
+          /*
+           * We only want to process 3 elements through the handler, but given buffering within the projections
+           * at-least-once impl. we actually process +1 element than we requested with the TestSink.probe.
+           *
+           * See https://github.com/akka/akka-projection/issues/462 for a possible solution.
+           */
           probe.request(2);
           probe.expectNextN(2);
           assertEquals("abc|def|ghi|", str.toString());

--- a/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
+++ b/akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java
@@ -182,7 +182,7 @@ public class JdbcProjectionTest extends JUnitSuite {
   @NotNull
   private JavaPartialFunction<TestSubscriber.SubscriberEvent, Object> expectErrorMessage(
       String msg) {
-    return new JavaPartialFunction<>() {
+    return new JavaPartialFunction<TestSubscriber.SubscriberEvent, Object>() {
       public String apply(TestSubscriber.SubscriberEvent in, boolean isCheck) {
         if (in instanceof TestSubscriber.OnError) {
           TestSubscriber.OnError err = (TestSubscriber.OnError) in;

